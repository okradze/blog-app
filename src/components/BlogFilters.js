import React, { useEffect } from 'react';
import { connect } from 'react-redux';
import { setTitleFilter, searchByAuthor, searchByTitle } from '../actions/filters';

export const BlogFilters = (props) => {
    const onTitleChange = (e) => {
        props.setTitleFilter(e.target.value);
    };

    const onSortChange = (e) => {
        if(e.target.value === 'title') {
            props.searchByTitle();
        } else if (e.target.value === 'author') {
            props.searchByAuthor();
        }
    };

    useEffect(() => {
        return () => {
            props.setTitleFilter('');
        }
    }, [])

    return (
        <div className="flex items-center">
            <input type="text" className="py-2 px-4 mr-4 rounded-sm" placeholder="Search by" value={props.filters.title} onChange={onTitleChange} />
            <select className="py-2 px-4 rounded-sm" value={props.filters.searchBy} onChange={onSortChange}>
                <option value="title">Title</option>
                <option value="author">Author</option>
            </select>
        </div>
    );
}

const mapStateToProps = (state) => ({
    filters: state.filters
});

const mapDispatchToProps = (dispatch) => ({
    setTitleFilter: (title) => dispatch(setTitleFilter(title)),
    searchByAuthor: () => dispatch(searchByAuthor()),
    searchByTitle: () => dispatch(searchByTitle())
});

export default connect(mapStateToProps, mapDispatchToProps)(BlogFilters);