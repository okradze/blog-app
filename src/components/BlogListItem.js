import React, { useState } from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';
import 'intersection-observer';

export const BlogListItem = ({ title, author, id, to, createdAt, url, photoURL }) => {
	const [ visibleCaption, setVisibleCaption ] = useState(true);

	return (
		<article className="blog-item">
			<div className="blog-info">
				<img src={photoURL} className="blog-info__photo" alt={author} />
				<div className="blog-info__content">
					<p className="blog-info__author">{author}</p>
					<time dateTime={new Date(createdAt)} className="blog-info__date">
						{moment(createdAt).format('MMM Do, YYYY')}
					</time>
				</div>
			</div>

			<figure className="blog-item__thumbnail">
				{visibleCaption && <div className="blog-item__thumbnail-no" />}
				<img onLoad={() => setVisibleCaption()} src={url} alt={title} className="blog-item__thumbnail-img" />
			</figure>

			<div className="blog-item__content">
				<h2 className="blog-item__title">{title}</h2>
				{to === 'edit' && (
					<Link className="read-more" to={`${to}/${id}`}>
						Edit Blog
					</Link>
				)}
				<Link className="read-more" to={`read/${id}`}>
					Read more...
				</Link>
			</div>
		</article>
	);
};
export default BlogListItem;
