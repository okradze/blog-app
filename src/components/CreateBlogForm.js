import React from 'react';
import { connect } from 'react-redux';
import moment from 'moment';
import { createBlog } from '../actions/blogs';

export class CreateBlogForm extends React.Component {
    state = {};
    onTitleChange = (e) => {
        const title = e.target.value;
        this.setState(() => ({ title }));
    };
    onBodyChange = (e) => {
        const body = e.target.value;
        this.setState(() => ({ body }));
    };
    onSubmit = (e) => {
        e.preventDefault();
        if(!this.state.title || !this.state.body){
            this.setState(() => ({ error: 'Please provide blog title and body' }));
        } else {
            this.setState(() => ({ error: '' }));
            this.props.createBlog({
                title: this.state.title,
                body: this.state.body,
                createdAt: moment().valueOf()
            });
        }
    };
    render() {
        return (
            <form>
                <input type="text" placeholder="Blog title" onChange={this.onTitleChange} />
                <textarea placeholder="Blog body" onChange={this.onBodyChange}></textarea>
                <button onClick={this.onSubmit}>Create Blog</button>
            </form>
        );
    }
};

const mapDispatchToProps = (dispatch) => ({
    createBlog: (blog) => dispatch(createBlog(blog))
});

export default connect(undefined, mapDispatchToProps)(CreateBlogForm);