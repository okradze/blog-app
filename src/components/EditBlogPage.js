import React from 'react';

export const EditBlogPage = (props) => (
    <div>Edit blog {props.match.params.id}</div>
);

export default EditBlogPage;