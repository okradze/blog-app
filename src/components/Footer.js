import React from 'react';
import { Link } from 'react-router-dom';

export const Footer = () => (
	<footer className="footer">
		<div className="container footer__content">
			<nav className="footer__nav">
				<Link className="footer__link" to="/dashboard">
					Dashboard
				</Link>
				<Link className="footer__link" to="/create">
					Create
				</Link>
				<Link className="footer__link" to="/read">
					Blogs
				</Link>
			</nav>
			<p className="footer__copy">Copyright &copy; 2019 Mirian Okradze</p>
		</div>
	</footer>
);

export default Footer;
