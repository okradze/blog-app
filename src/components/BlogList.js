import React from 'react';
import { connect } from 'react-redux';
import BlogListItem from './BlogListItem';
import selectBlogs from '../selectors/blogs';

export const BlogList = ({ blogs }) => (
    <div>
        {(
            blogs.map(blog => <BlogListItem key={blog.id} blog={blog} />)
        )}
    </div>
);

const mapStateToProps = (state) => ({
    blogs: selectBlogs(state.blogs, state.filters)
});

export default connect(mapStateToProps)(BlogList);