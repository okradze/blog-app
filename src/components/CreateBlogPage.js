import React, { useState } from 'react';
import CreateBlogForm from './CreateBlogForm';
import { startCreateBlog } from '../actions/blogs';

export const CreateBlogPage = ({ history }) => {
	const [ submitLoading, setSubmitLoading ] = useState(false);
	const [ submitDisabled, setSubmitDisabled ] = useState(false);

	const onSubmit = (blog, uid) => {
		setSubmitLoading(true);
		setSubmitDisabled(true);
		startCreateBlog(blog, uid).then(() => {
			history.push('/dashboard');
		});
	};
	return (
		<React.Fragment>
			<div className="container mb-auto">
				<CreateBlogForm submitDisabled={submitDisabled} submitLoading={submitLoading} onSubmit={onSubmit} />
			</div>
		</React.Fragment>
	);
};

export default CreateBlogPage;
