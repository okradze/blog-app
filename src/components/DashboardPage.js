import React from 'react';
import BlogFilters from './BlogFilters';
import BlogList from './BlogList';

export const DashboardPage = () => (
    <div className="dashboard-page container">
        <BlogFilters />
        <BlogList />
    </div>
);

export default DashboardPage;