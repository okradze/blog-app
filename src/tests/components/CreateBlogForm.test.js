import React from 'react';
import { shallow } from 'enzyme';
import { CreateBlogForm } from '../../components/CreateBlogForm';

let onSubmit, title, body, createdAt, submitDisabled, submitLoading, removeDisabled, removeLoading, onRemove, auth;

beforeEach(() => {
	onSubmit = jest.fn();
	title = 'title';
	body = 'body';
	createdAt = 100000;
	submitDisabled = false;
	submitLoading = false;
	removeDisabled = false;
	removeLoading = false;
	onRemove = jest.fn();
	auth = {
		uid: 'abc123',
		displayName: 'Mirina',
		photoURL: 'abc',
	};
});

test('should render component correctly', () => {
	const wrapper = shallow(
		<CreateBlogForm
			onSubmit={onSubmit}
			title={title}
			body={body}
			createdAt={createdAt}
			submitDisabled={submitDisabled}
			submitLoading={submitLoading}
			removeDisabled={removeDisabled}
			removeLoading={removeLoading}
			onRemove={onRemove}
			auth={auth}
		/>,
	);
	expect(wrapper).toMatchSnapshot();
});

test('should render component correctly when edit page', () => {
	const wrapper = shallow(
		<CreateBlogForm
			onSubmit={onSubmit}
			title={title}
			body={body}
			createdAt={createdAt}
			submitDisabled={submitDisabled}
			submitLoading={submitLoading}
			removeDisabled={removeDisabled}
			removeLoading={removeLoading}
			onRemove={onRemove}
			auth={auth}
			isEdit={true}
		/>,
	);
	expect(wrapper).toMatchSnapshot();
});

test('should call onSubmit on form submit', () => {
	const wrapper = shallow(
		<CreateBlogForm
			onSubmit={onSubmit}
			title={title}
			body={body}
			createdAt={createdAt}
			submitDisabled={submitDisabled}
			submitLoading={submitLoading}
			removeDisabled={removeDisabled}
			removeLoading={removeLoading}
			onRemove={onRemove}
			auth={auth}
		/>,
	);
	wrapper.find('form').simulate('submit', {
		preventDefault: () => {},
	});
	expect(onSubmit).toHaveBeenCalled();
});
