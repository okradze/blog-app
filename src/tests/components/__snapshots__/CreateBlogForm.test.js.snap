// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render component correctly 1`] = `
<form
  className="create-form"
  onSubmit={[Function]}
>
  <input
    className="input create-form__input"
    onChange={[Function]}
    placeholder="Blog title"
    type="text"
    value="title"
  />
  <input
    accept="image/*"
    className="input file-upload"
    multiple={true}
    onChange={[Function]}
    type="file"
  />
  <Quill
    body="body"
    setBody={[Function]}
  />
  <button
    className="button create-form__button"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    Save Blog
  </button>
</form>
`;

exports[`should render component correctly when edit page 1`] = `
<form
  className="create-form"
  onSubmit={[Function]}
>
  <input
    className="input create-form__input"
    onChange={[Function]}
    placeholder="Blog title"
    type="text"
    value="title"
  />
  <Quill
    body="body"
    setBody={[Function]}
  />
  <button
    className="button create-form__button"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    Save Blog
  </button>
  <button
    className="button button--grey"
    disabled={false}
    onClick={[MockFunction]}
    type="button"
  >
    Remove
  </button>
</form>
`;
